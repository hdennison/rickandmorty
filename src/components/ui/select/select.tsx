import { ChevronsUpDownIcon } from "../icon/icon";
import { cn } from "../utils";
import { cva, VariantProps } from "class-variance-authority";

// I strongly recommend to use 'Select' from 'radix-ui' instead of this custom select component
const selectVariants = cva(
  "flex relative items-center gap-2 border-[#aaa] rounded-md border-1 h-10 fake-input",
  {
    variants: {
      disabled: {
        true: "bg-[#eee] text-[#999]",
        false: "bg-[#fafafa]",
      },
    },
    defaultVariants: {
      disabled: false,
    },
  },
);

export interface PanelProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof selectVariants> {
  disabled?: boolean;
}

const Select = ({
  children,
  className,
  disabled = false,
  ...props
}: React.SelectHTMLAttributes<HTMLSelectElement>) => (
  <div className={cn(selectVariants({ disabled, className }))}>
    <select
      className="appearance-none flex-1 p-2 pl-3 pr-8 outline-none"
      disabled={disabled}
      {...props}
    >
      {children}
    </select>
    <ChevronsUpDownIcon className="absolute right-2" size={16} />
  </div>
);

const Option = ({
  children,
  ...props
}: React.OptionHTMLAttributes<HTMLOptionElement>) => (
  <option {...props}>{children}</option>
);

Select.Option = Option;

export { Select };
