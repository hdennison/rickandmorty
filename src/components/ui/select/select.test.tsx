import { expect, test } from "vitest";
import { render } from "vitest-browser-react";
import { Select } from "./select";

test("Button with icon", () => {
  const select = render(
    <Select>
      <Select.Option>Option 1</Select.Option>
      <Select.Option selected>Option 2</Select.Option>
    </Select>,
  );

  // Snapshots are not the best solution, it woould be much better to set up Storybook + Chromatic
  expect(select).toMatchSnapshot();
});
