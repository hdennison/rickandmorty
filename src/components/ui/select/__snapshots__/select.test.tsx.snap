// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button with icon 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    
    
    
  


    <div>
      <div
        class="flex relative items-center gap-2 border-[#aaa] rounded-md border-1 h-10 fake-input bg-[#fafafa]"
      >
        <select
          class="appearance-none flex-1 p-2 pl-3 pr-8 outline-none"
        >
          <option>
            Option 1
          </option>
          <option>
            Option 2
          </option>
        </select>
        <svg
          class="lucide lucide-chevrons-up-down absolute right-2"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 15 5 5 5-5"
          />
          <path
            d="m7 9 5-5 5 5"
          />
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex relative items-center gap-2 border-[#aaa] rounded-md border-1 h-10 fake-input bg-[#fafafa]"
    >
      <select
        class="appearance-none flex-1 p-2 pl-3 pr-8 outline-none"
      >
        <option>
          Option 1
        </option>
        <option>
          Option 2
        </option>
      </select>
      <svg
        class="lucide lucide-chevrons-up-down absolute right-2"
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m7 15 5 5 5-5"
        />
        <path
          d="m7 9 5-5 5 5"
        />
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholder": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
